File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class Trie {
private:
    struct TrieNode{
        TrieNode* children[26];
        bool isLeafNode;
    };
    TrieNode* root = new TrieNode();
public:
    Trie() {}
    
    TrieNode* getTrieNode(){
        TrieNode* node = new TrieNode();
        for (int i = 0; i < 26; ++i) node -> children[i] = NULL;
        node -> isLeafNode = false;
        return node;
    }
    
    void insert(string word) {
        if (search(word)) return;
        if (!root) root = getTrieNode();
        
        TrieNode* node = root;
        
        for (int i = 0; i < word.size(); ++i){
            int idx = word[i] - 'a';
            if (!node -> children[idx]){
                node -> children[idx] = getTrieNode();
            }
            node = node -> children[idx];
        }
        
        node -> isLeafNode = true;
    }
    
    bool search(string word) {
        TrieNode* node = root;
        
        if (!root) return false;
        
        for (int i = 0; (i < word.size()) && node; ++i){
            int idx = word[i] - 'a';
            node = node -> children[idx];
        }
        
        if (node) return node -> isLeafNode;
        return false;
    }
    
    bool startsWith(string prefix) {
        TrieNode* current_node = root;
        if (!root) return false;
        
        for (int i = 0; (i < prefix.size()) && current_node; ++i){
            int idx = prefix[i] - 'a';
            current_node = current_node -> children[idx];
        }
        
        return (current_node != NULL);
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class Node:
    def __init__(self, key: chr, data: str = None):
        self.key = key
        self.data = data
        self.children = {}

class Trie:

    def __init__(self):
        self.head = Node(None)

    def insert(self, word: str) -> None:
        current_node = self.head
        
        for char in word:
            if not char in current_node.children:
                current_node.children[char] = Node(char)
            current_node = current_node.children[char]
        current_node.data = word

    def search(self, word: str) -> bool:
        current_node = self.head
        
        for char in word:
            if char in current_node.children:
                current_node = current_node.children[char]
            else:
                return False
        
        if current_node.data:
            return True
        else:
            return False
            

    def startsWith(self, prefix: str) -> bool:
        current_node = self.head
        
        for char in prefix:
            if char in current_node.children:
                current_node = current_node.children[char]
            else:
                return False
            
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)